data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/ProcessWorker.java

```diff
@@ -44,41 +45,73 @@ public ProcessWorker(
+            while (!areBuffersEmpty()) {
+                doRun();
+            }
+            LOG.info("Processor shutdown phase 2 complete.");
+
+            // Phase 3 - execute until peer forwarder drain period expires (best effort to process all peer forwarder data)
+            final long drainTimeoutExpiration = System.currentTimeMillis() + pipeline.getPeerForwarderDrainTimeout().toMillis();
+            LOG.info("Beginning processor shutdown phase 3, iterating until {}.", drainTimeoutExpiration);
+            while (System.currentTimeMillis() < drainTimeoutExpiration) {
+                doRun();
+            }
+            LOG.info("Processor shutdown phase 3 complete.");
+
+            // Phase 4 - prepare processors for shutdown
+            LOG.info("Beginning processor shutdown phase 4, preparing processors for shutdown.");
+            processors.forEach(Processor::prepareForShutdown);
+            LOG.info("Processor shutdown phase 4 complete.");
+
+            // Phase 5 - execute until processors are ready to shutdown
+            LOG.info("Beginning processor shutdown phase 5, iterating until processors are ready to shutdown.");
+            while (!areComponentsReadyForShutdown()) {
+                doRun();
+            }
+            LOG.info("Processor shutdown phase 5 complete.");
         } catch (final Exception e) {
             LOG.error("Encountered exception during pipeline {} processing", pipeline.getName(), e);
         }
     }
 
-    /**
-     * Shutdown should be handled end to end.
-     *
-     * @return
-     */
-    private boolean shouldStop() {
-        return pipeline.isStopRequested() && areComponentsReadyForShutdown();
+    private void doRun() {
+        final Map.Entry<Collection, CheckpointState> readResult = readBuffer.read(pipeline.getReadBatchTimeoutInMillis());
+        Collection records = readResult.getKey();
+        final CheckpointState checkpointState = readResult.getValue();
+        //TODO Hacky way to avoid logging continuously - Will be removed as part of metrics implementation
+        if (records.isEmpty()) {
+            if(!isEmptyRecordsLogged) {
+                LOG.info(" {} Worker: No records received from buffer", pipeline.getName());
+                isEmptyRecordsLogged = true;
+            }
+        } else {
+            LOG.info(" {} Worker: Processing {} records from buffer", pipeline.getName(), records.size());
+        }
+        //Should Empty list from buffer should be sent to the processors? For now sending as the Stateful processors expects it.
+        for (final Processor processor : processors) {
+            records = processor.execute(records);
+        }
+        if (!records.isEmpty()) {
+            postToSink(records);
+        }
+        // Checkpoint the current batch read from the buffer after being processed by processors and sinks.
+        readBuffer.checkpoint(checkpointState);
+    }
+
+    private boolean areBuffersEmpty() {
```

Comment:
This is creating a strong coupling between peer-forwarding and the `ProcessWorker`. I'd like to see if we can do this only via the `Processor::isReadyForShutdown()` method.

If that won't work, then there seems an opportunity to use the `Buffer` in such a way that it is knowledgable of peer-forwarding (possibly using another decorator). Then this code doesn't need to change too much.

Reply:
The issue with having the logic in `isReadyForShutdown` is that `prepareForShutdown` is called first. For the AggregateProcessor, `prepareForShutdown` causes the existing data to be flushed before the group duration timeout.  For the ServiceMapStatefulProcessor, `prepareForShutdown` closes the current window, but opens a new one at the end so any subsequent data will remain in flux until the window duration ends, thereby making the `prepareForShutdown` call for nothing.

Essentially we want at most one more iteration of `doRun` to occur after `prepareForShutdown` is called. By moving the timeout/peerforwarder buffer empty check to `Processor::isReadyForShutdown()`, we are likely (all but guaranteed if the drain_timeout is non-zero) to get multiple iterations of `doRun` after `prepareForShutdown`

>there seems an opportunity to use the Buffer in such a way that it is knowledgable of peer-forwarding (possibly using another decorator)

Could you elaborate on this a bit? I think there could be a solution with a decorator that has the local buffer and the peer forwarder buffer as member variables. All but the `isEmpty()` method would be pass throughs to the local buffer and `isEmpty()` would be something like:
```
@Override
public boolean isEmpty() {
  return peerForwarderBuffer.isEmpty() && localBuffer.isEmpty();
}
```

I am not sure how much refactoring would be required to achieve the above since a PeerForwarderProvider constructs the `PeerForwarder` and it's buffer in the constructor of `PeerForwardingProcessorDecorator`

Reply:
> I think there could be a solution with a decorator that has the local buffer and the peer forwarder buffer as member variables.

Yes, this is very close to what I was suggesting. We'd have to update the `PipelineParser` to decorate the `Buffer` object that it normally creates.

> since a PeerForwarderProvider constructs the PeerForwarder and it's buffer in the constructor of PeerForwardingProcessorDecorator

I think the `PeerForwarderProvider` could be able to supply all the buffers for a given pipeline. Then we can check each of these buffers.

Reply:
Ok, I'll look into refactoring this